api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,273 @@
+    // the "request_body" field set.
+    BodyResponse request_body = 3;
+
+    // The server must send back this message in response to a message with
+    // the "response_body" field set.
+    BodyResponse response_body = 4;
+
+    // The server must send back this message in response to a message with
+    // the "request_trailers" field set.
+    TrailersResponse request_trailers = 5;
+
+    // The server must send back this message in response to a message with
+    // the "response_trailers" field set.
+    TrailersResponse response_trailers = 6;
+
+    // If specified, return the specified response downstream to the client
+    // and end further filter processing on the request path.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+message HeadersResponse {
```

Comment:
I think this message you also want:
1) To be able to specify trailers
2) To be able to specify end_stream

This would effectively allow you to create a complete request/response given just the headers if you want and override everything else that might be coming.

Reply:
Yes ^ is exactly right. It's something that can be done by existing C++ filters and is something that ends up being done occasionally. Essentially, what I'm talking about is transformations in which we do things like:
1) Make a header only request into a not header only request, and optionally add data and trailers at the same time.
2) Make a not header only request into a header only request.

The other major use case that you need to think through is something like the following:
1) ext proc server gets header only request.
2) respond and make not header only request, but don't supply enough info (body, trailers, etc.) to complete the request.
3) ext proc responds and streams body/trailer data as it becomes available.

These types of things are all things that can be done with C++ filters. Now, I don't think it's required that we support all of this on day 1, but I suggest making sure the design *can* support all of these things.

The TL/DR request is this: Study filter.h, all of the iteration return codes, and all of the async completion possibilities, and make sure they could be modeled by this API. I think this may involve a protocol that is not always request/response. It may be a request followed by a series of streamed responses for example.